### PR TITLE
feat(Devtools): Support for persist, lock, pause

### LINKED ISSUE #853: RFC: Dev Tools support for persist, lock, pause
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
</code></pre>

## What is the current behavior?
The redux dev tools extension has the following features which are not usable with ngrx at the moment:
- persist: If enabled, the state of the redux dev tools survive a full reload of the page. The persisted actions are fired against the initial state. This makes developing applications faster. Imagine you don't use HMR and have to fix something from the example application - every reload throws you back to the login page. With the persist feature, this would no longer be the case.
- lock/unlock: If locked, no new actions will get through to the reducer, the app is basically frozen. In combination with persist / HMR, this is a great developer experience when you want to focus on a specific feature with a specific number/order of actions.
- pause/unpause: If paused, the redux dev tools will not record the actions fired. The app will continue to work. Useful if you don't want to get spammed with lots of actions. When you unpause, the current state will be committed and recording is resumed.

Locking and pausing is explained in more details [here](https://medium.com/@zalmoxis/improve-your-development-workflow-with-redux-devtools-extension-f0379227ff83).

## Expected behavior:
Features also work with ngrx dev tools.

## Other information:
I'm willing to help with PRs.

### PR DESCRIPTION
**Pull request by dummdidumm**
Closes #853:

Devtools now support:
- persist: When devtools option is clicked, state survives full page reload
- lock: When activated, all actions are ignored
- pause: When activated, recording of actions is paused

Fixes part of #919:

When importing a state, there is now a one second delay to give lazy loaded reducers time to instantiate.
Think of this situation: User is on a page where a reducer was lazy loaded. He wants to use the persist feature. The page is reloaded. Now the "import state" action is fired from the devtools, but the actions are replayed too soon because the lazy loaded reducer is not instantiated yet. That means, the actions are replayed with incomplete reducers. To fix this, a small delay is added before importing a state.

### CODE REVIEW

File: modules/router-store/src/router_store_module.ts [hunk 1]
```diff
@@ -279,7 +279,10 @@ export class StoreRouterConnectingModule {
 
     if (this.router.url !== this.storeState[this.stateKey].state.url) {
       this.navigationTriggeredByDispatch = true;
-      this.router.navigateByUrl(this.storeState[this.stateKey].state.url);
```
Comment (brandonroberts) [inline modules/router-store/src/router_store_module.ts]: This should be a separate PR. I don't think this will work for SSR apps also because you can't guarantee this will be finished before the next navigation starts.

**Comment by dummdidumm** [comment]
I removed the router-store-commit from the PR and instead opened #960 .

What do you think about the other changes?

**Comment by dummdidumm** [comment]
@brandonroberts any plans if/when this will be merged?

**Comment by brandonroberts** [comment]
I'll take a look this week. I'd like something a little more predictable for waiting for the reducers to load when importing, but we'll see.

**Comment by dummdidumm** [comment]
Thanks for the update!

Regarding the import and "wait on lazy loaded reducers": Since lazy loaded features/reducers are directly tied to lazy loaded modules - what about injecting the `Router` with `@Optional()` and then on import
- check if the router exists
- if yes, if it did navigate yet
- if it did not navigate yet, it means the import is done straight after reload, so we wait for the initial router navigation to end (reducers will be loaded by then) and then import the state. 
- if it did already navigate, we can be sure that every module/reducer we know of at this point is already loaded, so we import the state straight away.
- if no router dependency exists, we fall back to waiting for 1 second. Maybe users use a different router implementation or something like that. 

I could add a commit implementing this behavior. What do you think?

Side note: This would not solve all import issues with lazy loading. What about a state history which does some navigation action and only at this point a new reducer is loaded (#919)? But I think this is far more complex to think through / implement and does not belong into this PR.

**Comment by dummdidumm** [comment]
@brandonroberts any updates on this? What do you think about the more predictable "wait for reducers by using Router"-proposal?

**Comment by dummdidumm** [comment]
Sorry to ask again but is there any timeline when (or if at all) this will be merged? Any feedback would be appreciated.

**Comment by brandonroberts** [comment]
Apologies for getting behind on this one. Will you rebase against master?

**Comment by dummdidumm** [comment]
Thanks for the update, I will do it next week.

**Comment by dummdidumm** [comment]
I resolved some new merge conflicts. The build fails but this seems to have nothing to do with the commits (bazel ran out of memory and crashed).

File: modules/store-devtools/src/extension.ts [hunk 2]
```diff
@@ -163,7 +169,31 @@ export class DevtoolsExtension {
     // Listen for lifted actions
     const liftedActions$ = changes$.pipe(
       filter(change => change.type === ExtensionActionTypes.DISPATCH),
-      map(change => this.unwrapAction(change.payload))
+      map(change => this.unwrapAction(change.payload)),
+      concatMap((action: any) => {
+        if (action.type === IMPORT_STATE) {
+          if (this.router) {
```
Comment (brandonroberts) [inline modules/store-devtools/src/extension.ts]: I don't think the router should be integrated here. The route may need the state we are importing, but this prevents that from happening until the navigation is over.

File: modules/store-devtools/src/extension.ts [hunk 3]
```diff
@@ -185,9 +215,8 @@ export class DevtoolsExtension {
     return typeof action === 'string' ? eval(`(${action})`) : action;
   }
 
-  private getExtensionConfig(instanceId: string, config: StoreDevtoolsConfig) {
+  private getExtensionConfig(config: StoreDevtoolsConfig) {
     const extensionOptions: ReduxDevtoolsExtensionConfig = {
-      instanceId: instanceId,
```
Comment (brandonroberts) [inline modules/store-devtools/src/extension.ts]: Why is this removed?

File: modules/store-devtools/src/extension.ts (hunk 2 above)
Comment (brandonroberts) [reply modules/store-devtools/src/extension.ts]: Maybe we can listen for any `@ngrx/store/update-reducers` actions and delay the import until none have been received within a span of time.

**Comment by brandonroberts** [comment]
@dummdidumm we are still working through the Bazel memory issues, but at long as the tests pass, its fine for now. I left you some feedback on the PR. Let me know what you think.

File: modules/store-devtools/src/extension.ts (hunk 2 above)
Comment (dummdidumm) [reply modules/store-devtools/src/extension.ts]: `liftedActions$` retrieves the IMPORT_STATE-action before it is passed to the reducer. So as long as IMPORT_STATE is not emitted by `liftedActions$`, the import will not be triggered. `update-reducers` etc will be fired before the route is loaded because that happens on a different stream of actions (`actions$`), so there is no block. So in my opinion this is fine.

**Comment by dummdidumm** [comment]
I added you some feedback, let me know what you think.

File: modules/store-devtools/src/extension.ts (hunk 2 above)
Comment (dummdidumm) [reply modules/store-devtools/src/extension.ts]: I think it's not quite clear what I mean, so I'll expand on it:

### Situation 1: Router is used by user. User clicks the "import"-button of the redux devtools and manually imports a state
The added logic will not matter because  router.navigated will be true (the first navigation already happened) , therefore nothing to wait on, therefore the IMPORT action is fired straight away.

### Situation 2: Router is used by user. User turned on the "persist"-feature and reloads.
The added logic now matters. router.navigated will be false because the router is in the process of navigating when the action comes in. What that means is that the modules connected to that route are still getting instantiated, so for example components/reducers/effects for that route are not there yet.  So if we just let the IMPORT action pass through, the reducer would import the state too soon, some actions would get ignored because the reducers handling it do not exist yet. That's why we wait until the NavigationEnd-event of the router is fired. At this point we know all the initialization logic is done.<br>
There will be no blocking of state needed for the router because initial reducers/effects/state will be loaded during navigation. These things do not come through the `liftedActions$` stream, so blocking further events there until we are finished importing is not an issue. Also something like "oh you are not logged in, I will redirect you" is fine because, well, we just import the state after that and the application will get back to the correct route.

### Situation 3: Router is not used by user.
The whole router logic does not matter now, we wait 1 second just to make sure.

Comment (brandonroberts) [reply modules/store-devtools/src/extension.ts]: I understand what's going on here, but I'd still rather support this in a consistent way without the router if possible.

Idea: After the IMPORT_STATE action comes in, wait some amount of time, then switch to listening for update-actions and debounce the IMPORT_STATE action. We can inject and listen to the store actions being dispatched.

Comment (dummdidumm) [reply modules/store-devtools/src/extension.ts]: Like this?

```
concatMap((action: any) => {
  if(action.type === IMPORT_STATE) {
    return actions$.pipe(
      ofType(UPDATE), // '@ngrx/store/update-reducers'
      debounceTime(1000),
      map(() => action)
    );
  }
...
```

Comment (dummdidumm) [reply modules/store-devtools/src/extension.ts]: We cannot know whether the IMPORT_STATE action is fired because of the persist option turned on or because of a manual import. But that's no harm, the user just has a minimal delay (of 1 second) until the manual import starts.

Comment (dummdidumm) [reply modules/store-devtools/src/extension.ts]: Good point, a race condition / timeout in the likes of "there is no UPDATE action at all" that fires after 1 second is needed.
I will implement it and update the PR.

File: modules/store-devtools/src/extension.ts [hunk 4]
```diff
@@ -17,6 +28,8 @@ import {
   sanitizeStates,
   unliftState,
 } from './utils';
+import { UPDATE } from '../../store/src';
```
Comment (brandonroberts) [inline modules/store-devtools/src/extension.ts]: This should be `@ngrx/store` instead of a relative import
